cli/deployment/api_client.go

```diff
@@ -11,73 +11,233 @@ import (
 )
 
+type Status int
+
+const (
+	Done     Status = 0
+	Pending  Status = 1
+	Running  Status = 2
+	NotFound Status = 3
+	Error    Status = -1
+)
+
+const (
+	ActionApply  string = "apply"
+	ActionDelete string = "delete"
+	ActionCreate string = "create"
+	ActionUpdate string = "update"
+)
+
+var actions = []string{ActionApply, ActionDelete, ActionCreate, ActionUpdate}
+
+// Function runGCloud exposed to variable in order to mock it inside api_client_test.go
 // The runGCloud function runs the gcloud tool with the specified arguments. It is implemented
 // as a variable so that it can be mocked in tests of its exported consumers.
 var runGCloud = func(args ...string) (result string, err error) {
-	args = append(args, "--format", "yaml")
 	log.Println("gcloud", strings.Join(args, " "))
 	cmd := exec.Command("gcloud", args...)
-	outBuff := &bytes.Buffer{}
-	errBuff := &bytes.Buffer{}
-	cmd.Stdout = outBuff
-	cmd.Stderr = errBuff
 	// pass user's PATH env variable, expected gcloud executable can be found in PATH
 	cmd.Env = append(cmd.Env, fmt.Sprintf("PATH=%s", os.Getenv("PATH")))
 
-	if err := cmd.Start(); err != nil {
-		log.Printf("Failed to start cmd: %v, \n Output:\n %v", err, errBuff.String())
-		return "", err
-	}
+	output, err := cmd.CombinedOutput()
 
-	if err := cmd.Wait(); err != nil {
-		log.Printf("Cmd returned error: %v, \n Output:\n %v", err, errBuff.String())
-		return errBuff.String(), err
+	if err != nil {
+		log.Printf("failed to start cmd: %v, output: %v", err, string(output))
+		return string(output), err
 	}
 
-	return outBuff.String(), err
+	return string(output), err
 }
 
 // GetOutputs retrive existing Deployment outputs using gcloud and store result in map[string]string
 // where "resourceName.propertyName" is key, and value is string representation of the output value.
-func GetOutputs(name string, project string) (map[string]string, error) {
-	data, err := runGCloud("deployment-manager", "manifests", "describe", "--deployment", name, "--project", project)
+func GetOutputs(project string, name string) (map[string]interface{}, error) {
+	data, err := runGCloud("deployment-manager", "manifests", "describe", "--deployment", name, "--project", project, "--format", "yaml")
 	if err != nil {
 		log.Printf("Failed to get deployment manifest: %v", err)
 		return nil, err
 	}
 	return parseOutputs(data)
 }
 
-// Create creates a new Deployment based on a Deployment object passed into it.
-func Create(deployment *Deployment) (*Deployment, error) {
+/*
+returns project id taken from local gcloud configuration
+*/
+func GCloudDefaultProjectID() (string, error) {
+	data, err := runGCloud("config", "list", "--format", "yaml")
+	if err != nil {
+		return "", err
+	}
+	out := struct {
+		Core struct {
+			Project string
+		}
+	}{}
+	err = yaml.Unmarshal([]byte(data), &out)
+	if err != nil {
+		return "", err
+	}
+	return out.Core.Project, nil
+}
+
+// Create deployment based on passed Deployment object passed into it.
```

Comment:
Per https://golang.org/doc/effective_go.html#commentary, doc comments for exported functions should generally be of the form "CreateOrUpdate creates a new deployment based on the Deployment object passed into it..." Since the function name has changed, this sentence should be changed to match.

Since the function creates or updates based on what the user has supplied, this might get confusing. I'd recommend making separate `Create` and `Update` functions that call this function, and making this one unexported. Then the doc comments for Create and Update are clear and the API is cleaner.